docker/join_layers.py

```diff
@@ -35,65 +36,105 @@
+               blobsum_to_unzipped, blobsum_to_zipped, blobsum_to_legacy):
+    self._config = config_file
+    self._blobsum_to_unzipped = blobsum_to_unzipped
+    self._blobsum_to_zipped = blobsum_to_zipped
+    self._blobsum_to_legacy = blobsum_to_legacy
+    config = json.loads(self._config)
+    self._manifest = json.dumps({
+        'schemaVersion': 2,
+        'mediaType': docker_http.MANIFEST_SCHEMA2_MIME,
+        'config': {
+            'mediaType': docker_http.CONFIG_JSON_MIME,
+            'size': len(self.config_file()),
+            'digest': 'sha256:' + hashlib.sha256(self.config_file()).hexdigest()
+        },
+        'layers': [
+            {
+                'mediaType': docker_http.LAYER_MIME,
+                'size': self.blob_size(diffid_to_blobsum[diff_id]),
+                'digest': diffid_to_blobsum[diff_id]
+            }
+            for diff_id in config['rootfs']['diff_ids']
+        ]
+    }, sort_keys=True)
 
   def manifest(self):
     """Override."""
-    raise Exception('This image only exists to expose config_file().')
+    return self._manifest
 
   def config_file(self):
     """Override."""
-    return self._cfg
+    return self._config
 
+  # Could be large, do not memoize
+  def uncompressed_blob(self, digest):
+    """Override."""
+    if digest not in self._blobsum_to_unzipped:
+      return self._blobsum_to_legacy[digest].uncompressed_blob(digest)
+    with open(self._blobsum_to_unzipped[digest], 'r') as reader:
+      return reader.read()
+
+  # Could be large, do not memoize
   def blob(self, digest):
     """Override."""
-    raise Exception('This image only exists to expose config_file().')
+    if digest not in self._blobsum_to_zipped:
+      return self._blobsum_to_legacy[digest].blob(digest)
+    with open(self._blobsum_to_zipped[digest], 'r') as reader:
+      return reader.read()
+
+  def blob_size(self, digest):
+    """Override."""
+    if digest not in self._blobsum_to_zipped:
+      return self._blobsum_to_legacy[digest].blob_size(digest)
+    info = os.stat(self._blobsum_to_zipped[digest])
+    return info.st_size
 
   # __enter__ and __exit__ allow use as a context manager.
   def __enter__(self):
-    """Open the image for reading."""
     return self
 
   def __exit__(self, unused_type, unused_value, unused_traceback):
-    """Close the image."""
     pass
 
 
-def create_image(output, layers, tag_to_layer=None, layer_to_tags=None):
+def create_bundle(output, tag_to_config, diffid_to_blobsum,
+                  blobsum_to_unzipped, blobsum_to_zipped, blobsum_to_legacy):
   """Creates a Docker image from a list of layers.
 
   Args:
     output: the name of the docker image file to create.
     layers: the layers (tar files) to join to the image.
-    tag_to_layer: a map from docker_name.Tag to the layer id it references.
+    tag_to_layer: a map from google3.third_party.bazel_rules.rules_docker.docker_name.Tag to the layer id it references.
```

Comment:
Clearly I built this elsewhere.  The funniest bit is that this rewrite (by the import tool) isn't even correct :)